tensorflow_model_optimization/python/core/sparsity_tf2/lthpruner_test.py

```diff
@@ -0,0 +1,486 @@
+    weight = tf.Variable(np.linspace(1.0, 100.0, 100), name="weights")
+    weight_dtype = weight.dtype.base_dtype
+    mask = tf.Variable(
+        tf.ones(weight.get_shape(), dtype=weight_dtype),
+        name="mask",
+        dtype=weight_dtype)
+    threshold = tf.Variable(
+        tf.zeros([], dtype=weight_dtype), name="threshold", dtype=weight_dtype)
+    pruning_vars = [(weight, mask, threshold)]
+
+    def linear_sparsity(step):
+      sparsity_val = tf.convert_to_tensor(
+          [0.0, 0.1, 0.1, 0.3, 0.3, 0.5, 0.5, 0.5, 0.5, 0.5])
+      return tf.convert_to_tensor(True), sparsity_val[step]
+
+    def weight_mask_op(pruning_vars):
+      values_and_vars = []
+      for weight, mask, _ in  pruning_vars:
+        weight.assign(tf.math.multiply(weight, mask))
+
+    # Set up pruning
+    p = pruner.LTHPruner(
+        pruning_schedule=linear_sparsity,
+        save_iteration=self.save_init,
+        block_size=self.block_size,
+        block_pooling_type=self.block_pooling_type)
+
+    optimizer = self.dummy_optimizer
+    optimizer.iterations.assign(0)
+    p.create_slots(optimizer, weight)
+    step = optimizer.iterations
+
+    non_zero_count = []
+    for _ in range(10):
+      step = optimizer.iterations
+      p.update_masks(pruning_vars, step)
+      weight_mask_op(pruning_vars)
+      optimizer.iterations.assign_add(1)
+
+      non_zero_count.append(tf.math.count_nonzero(weight))
+
+    # Weights pruned at steps 1,3,5
+    expected_non_zero_count = [100, 90, 90, 70, 70, 50, 50, 50, 50, 50]
+    self.assertAllEqual(expected_non_zero_count, non_zero_count)
+
+
+  def testSaveOriginalInitializations(self):
+    weight = tf.Variable(np.linspace(1.0, 100.0, 100), name="weights")
+    weight_dtype = weight.dtype.base_dtype
+    save_round = 0
+    n_rounds = 5
+    end_iter = 100
+    frequency, prune_ratio_per_round = get_lth_sparsity(save_round, n_rounds, self.target_sparsity, end_iter)
+    pruning_schedule = make_pruning_schedule(prune_ratio_per_round, 0, end_iter, frequency)
+    
+    p = pruner.LTHPruner(
+        pruning_schedule=pruning_schedule,
+        save_iteration=self.save_init,
+        block_size=self.block_size,
+        block_pooling_type=self.block_pooling_type)
+
+    optimizer = self.dummy_optimizer
+    optimizer.iterations.assign(0)
+    init_weights_before_pruning = weight.read_value()
+
+    def _train():
+      p.create_slots(optimizer, weight)
+      p.preprocess_weights(optimizer, weight, self.grad(weight))
+      weight.assign(tf.math.add(weight, sample_noise(0))) # update weights
+      p.postprocess_weights(optimizer, weight, self.grad(weight))
+      optimizer.iterations.assign_add(1)
+      initialization_slot = optimizer.get_slot(weight, "original_initialization")
+      return initialization_slot
+    
+    initialization_slot = _train()
+    self.assertAllEqual(initialization_slot, init_weights_before_pruning)
+    
+    initialization_slot = tf.function(_train)()
```

Comment:
Make sure to re-initialize your weights and optimizer between these.

Reply:
Create a brand-new optimizer object, and make _train() take an optimizer as an argument.